pkg/validation/interfaces/validator.go

```diff
@@ -0,0 +1,92 @@
+package validator
+
+import (
+	"sync"
+
+	"github.com/operator-framework/api/pkg/validation/errors"
+)
+
+// ValidatorFunc returns a ManifestResult containing errors and warnings from
+// validating some object. These are typically closures or methods.
+type ValidatorFunc func() errors.ManifestResult
+
+// ValidationFuncs is a set of validation functions.
+type ValidatorFuncs []ValidatorFunc
+
+// Validator is an interface for validating arbitrary objects. A Validator
+// returns a set of functions that each validate some underlying object.
+// This allows the implementer to easily break down each validation step
+// into discrete units if they wish, and perform either lazy or active
+// validation.
+type Validator interface {
```

Comment:
Some thoughts on the structure here. Could `Validator` itself be the thing that does a validation?

```golang
type Validator interface {
   Validate(objs ...interface{}) (results ...errors.ManifestResult)
}
```

Then `ValidatorFunc` could be a function type that implements `Validator`.

```golang
type ValidatorFunc func(objs ...interface{}) (...errors.ManifestResult)

func (v ValidatorFunc) Validate(objs ...interface{}) (results ...errors.ManifestResult) {
    return v(objs...)
} 
```

`Validators` can still be a list of validators, but can also implement `Validate` and aggregate the result.

```golang
type Validators []Validator

func (v Validators) Validate(objs ...interface{}) (results ...errors.ManifestResult) {
    for validate := range v { 
        results = append(results, validate(objs...)...)
    }

    return
}

```

Reply:
That's what I was originally thinking because it's simple. The interface usage isn't _that_ much different either:

```go
type customValidator struct{}

func (v customValidator) GetFuncs(objs ...interface{}) (funcs interfaces.ValidatorFuncs) {
	for _, obj := range objs {
		...
	}
	return funcs
}

func (v customValidator) Validate(objs ...interface{}) (results []errors.ManifestResult) {
	for _, obj := range objs {
		...
	}
	return results
}

func tester() {
	objs := []interface{}{...}
	// Validator with GetFuncs
	results := DefaultValidators(customValidator{}).Apply(objs)

	// Validator with Validate
	validators := append(DefaultValidators(), customValidator{})
	results := validators.Validate(objs)
}
```

However, the `GetFuncs()` method is more parallelizable, and we'll get a better speedup assuming the number of validators N << number of objects M, and that validation is a resource-intensive tast. That being said, I don't think the speedup will be large because N and M won't be more than one order of magnitude different in most cases. And it won't matter at all if objects are being validated actively (not lazily).

In the end I'm fine with either interface layout, since they're not too different and provide the same features.

Reply:
> However, the GetFuncs() method is more parallelizable

Wouldn't they be equivalently parallelizable? You could define the `Validators.Validate` method in such a way that it parallelizes calls to its validators. Alternatively, you could define a new `ParallelValidator` that has `Validators` as a field and fires its elements off in parallel instead of calling its `Validate` method.

Reply:
Going with the simpler, non-parallelized approach. We can re-assess later if we feel improvements are needed.